app/Policies/AlbumPolicy.php

```diff
@@ -119,31 +104,64 @@ public function canAccess(?User $user, ?AbstractAlbum $album): bool
 				throw LycheeAssertionError::createFromUnexpectedException($e);
 			}
 		} elseif ($album instanceof BaseSmartAlbum) {
-			return $this->isVisible($user, $album);
+			return $this->canSee($user, $album);
 		} else {
 			// Should never happen
 			return false;
 		}
 	}
 
 	/**
-	 * Check if an album is downloadable.
+	 * Check if current user can download the album.
 	 *
-	 * @param User|null      $user
-	 * @param BaseAlbum|null $baseAlbum
+	 * @param User|null          $user
+	 * @param AbstractAlbum|null $abstractAlbum
 	 *
 	 * @return bool
 	 *
 	 * @throws ConfigurationKeyMissingException
 	 */
-	public function canDownload(?User $user, ?BaseAlbum $baseAlbum): bool
+	public function canDownload(?User $user, ?AbstractAlbum $abstractAlbum): bool
 	{
-		if ($baseAlbum === null) {
-			return Configs::getValueAsBool('downloadable');
+		$default = Configs::getValueAsBool('downloadable');
+		// The root album always uses the global setting
+		if ($abstractAlbum === null) {
+			return $default;
+		}
+
+		if ($abstractAlbum instanceof BaseSmartAlbum) {
+			return $user !== null || $abstractAlbum->grants_download;
+		}
+
+		// TODO: when download rights are assigned to albums, we add more logic can be added here.
+		if ($abstractAlbum instanceof BaseAlbum) {
+			return $this->isOwner($user, $abstractAlbum) ||
+				$abstractAlbum->grants_download ||
+				($abstractAlbum->shared_with()->where('user_id', '=', $user?->id)->count() > 0 && $default);
```

Comment:
>  Why do we (suddenly) need this potentially costly condition? I

This costly condition was always there, it is just that it was hidden in the Album/getArchive code.

Reply:
> Consider an album with a lot of sub albums, i.e. in the thousands, or consider the case when we send the whole album tree to the frontend. Now, we perform on extra SQL query for each album.

Wondering if we should just pre-fetch the shared-with relation, this would lower the number of queries then.